src/surface/manifold_surface_mesh.cpp

```diff
@@ -1037,6 +1037,242 @@ Vertex ManifoldSurfaceMesh::insertVertex(Face fIn) {
+        toReassignVertex.push_back(he);
+      }
+      for (Halfedge he : toReassignVertex) {
+        heVertexArr[he.getIndex()] = vB.getIndex();
+      }
+
+      heNextArr[heD0.getIndex()] = heB2.getIndex();
+      heNextArr[heB2.getIndex()] = heD2.getIndex();
+      heNextArr[heC1.getIndex()] = heA1.getIndex();
+      heNextArr[heA1.getIndex()] = heC0.getIndex();
+      heFaceArr[heB2.getIndex()] = fD.getIndex();
+      heFaceArr[heA1.getIndex()] = fC.getIndex();
+      fHalfedgeArr[fC.getIndex()] = heC0.getIndex();
+      fHalfedgeArr[fD.getIndex()] = heD0.getIndex();
+      // boundary vertex already safe with boundary edge
+      if (!vB.isBoundary()) {
+        vHalfedgeArr[vB.getIndex()] = heA1.getIndex();
+      }
+      if (!vC.isBoundary()) {
+        vHalfedgeArr[vC.getIndex()] = heC0.getIndex();
+      }
+      if (!vD.isBoundary()) {
+        vHalfedgeArr[vD.getIndex()] = heB2.getIndex();
+      }
+
+      // Actually delete
+      deleteEdgeBundle(e);
+      deleteEdgeBundle(heB1.edge());
+      deleteEdgeBundle(heA2.edge());
+      deleteElement(vA);
+      deleteElement(fA);
+      deleteElement(fB);
+
+      v = vB;
+      return true;
+    }
+
+    else if (heA0.vertex().degree() == 3) {
+      Halfedge heA1 = heA0.next();
+      Halfedge heA2 = heA1.next();
+      Halfedge heB0 = heA0.twin();
+      Halfedge heB1 = heB0.next();
+      Halfedge heB2 = heB1.next();
+      Halfedge heC2 = heA2.twin();
+      Halfedge heC0 = heC2.next();
+      Halfedge heC1 = heC0.next();
+
+      Face fA = heA0.face();
+      Face fB = heB0.face();
+      Face fC = heC0.face();
+
+      Vertex vA = heA0.vertex();
+      Vertex vB = heB0.vertex();
+      Vertex vC = heC0.vertex();
+      Vertex vD = heC1.vertex();
+
+      heNextArr[heA1.getIndex()] = heC0.getIndex();
+      heNextArr[heC0.getIndex()] = heB2.getIndex();
+      heNextArr[heB2.getIndex()] = heA1.getIndex();
+      heFaceArr[heB2.getIndex()] = fC.getIndex();
+      heFaceArr[heA1.getIndex()] = fC.getIndex();
+      fHalfedgeArr[fC.getIndex()] = heC0.getIndex();
+      // boundary vertex already safe with boundary edge
+      if (!vB.isBoundary()) {
+        vHalfedgeArr[vB.getIndex()] = heA1.getIndex();
+      }
+      if (!vC.isBoundary()) {
+        vHalfedgeArr[vC.getIndex()] = heC0.getIndex();
+      }
+      if (!vD.isBoundary()) {
+        vHalfedgeArr[vD.getIndex()] = heB2.getIndex();
+      }
+
+      // Actually delete
+      deleteEdgeBundle(e);
+      deleteEdgeBundle(heB1.edge());
+      deleteEdgeBundle(heA2.edge());
+      deleteElement(vA);
+      deleteElement(fA);
+      deleteElement(fB);
+
+      v = vB;
+      return true;
+    }
+
+    else {
+      // return heA0.vertex();
+      throw std::runtime_error("what is this vertex " + std::to_string(heA0.vertex()) + " with degree " +
+                               std::to_string(heA0.vertex().degree()) + " on edge " + std::to_string(heA0.edge()) +
+                               "?\n");
```

Comment:
could this please be changed to have a simple informative message in the error?

Reply:
Yes -- I believe this is the case when an interior vertex has degree < 3, so I've just added a message saying that this case isn't supported for collapses.